sycamore/llms/openai.py

```diff
@@ -0,0 +1,252 @@
+
+            if self.api_version is None:
+                raise ValueError("api_version must be specified for Azure clients.")
+
+            return AzureOpenAIClient(
+                azure_endpoint=self.azure_endpoint,
+                azure_deployment=self.azure_deployment,
+                api_version=self.api_version,
+                api_key=self.api_key,
+                azure_ad_token=self.azure_ad_token,
+                azure_ad_token_provider=self.azure_ad_token_provider,
+                organization=self.organization,
+                max_retries=self.max_retries,
+                **self.extra_kwargs,
+            )
+
+        else:
+            raise ValueError(f"Invalid client_type {self.client_type}")
+
+    def get_guidance_model(self, model) -> Model:
+        if self.client_type == OpenAIClientType.OPENAI:
+            return GuidanceOpenAI(
+                model=model.name,
+                api_key=self.api_key,
+                organization=self.organization,
+                base_url=self.base_url,
+                max_retries=self.max_retries,
+                **self.extra_kwargs,
+            )
+        elif self.client_type == OpenAIClientType.AZURE:
+            # Note: Theoretically the Guidance library automatically determines which
+            # subclass to use, but this appears to be buggy and relies on a bunch of
+            # specific assumptions about how deployed models are named that don't work
+            # well with Azure. This is the only way I was able to get it to work
+            # reliably.
+            if model.is_chat:
+                cls = AzureOpenAIChat
+            else:
+                cls = AzureOpenAIInstruct
+
+            return cls(
+                model=model.name,
+                azure_endpoint=self.azure_endpoint,
+                azure_ad_token_provider=self.azure_ad_token_provider,
+                api_key=self.api_key,
+                version=self.api_version,
+                azure_deployment=self.azure_deployment,
+                azure_ad_token=self.azure_ad_token,
+                organization=self.organization,
+                max_retries=self.max_retries,
+                **self.extra_kwargs,
+            )
+
+        else:
+            raise ValueError(f"Invalid client_type {self.client_type}")
+
+
+# Allow rough backwards compatibility
+OpenAIClientParameters = OpenAIClientWrapper
+
+
+class OpenAI(LLM):
+    def __init__(
+        self,
+        model_name: Union[OpenAIModels, OpenAIModel, str],
+        api_key=None,
+        client_wrapper: Optional[OpenAIClientWrapper] = None,
+        params: Optional[OpenAIClientParameters] = None,
+        **kwargs,
+    ):
+        if isinstance(model_name, OpenAIModels):
+            self.model = model_name.value
+        elif isinstance(model_name, OpenAIModel):
+            self.model = model_name
+        elif isinstance(model_name, str):
+            self.model = OpenAIModels.from_name(model_name).value
```

Comment:
do we need to fail here if doesn't match?

Reply:
Good catch.